sdk/error.go

```diff
@@ -454,77 +456,211 @@ var errorsLanguages = []map[int]string{
 	errorsFrench,
 }
 
-// NewError just set an error with a root cause
-func NewError(target Error, root error) Error {
-	target.Root = root
-	return target
+// Error type.
+type Error struct {
+	ID         int    `json:"id"`
+	Status     int    `json:"-"`
+	Message    string `json:"message"`
+	UUID       string `json:"uuid,omitempty"`
+	StackTrace string `json:"stack_trace,omitempty"`
 }
 
-// WrapError constructs a stack of errors, adding context to the preceding error.
-func WrapError(err error, format string, args ...interface{}) error {
-	return errors.Wrap(err, fmt.Sprintf(format, args...))
+type errorWithStack struct {
+	root      error  // root error should be wrapped with stack
+	stack     *stack // used to generate inline call stack
+	httpError Error
 }
 
-// SetError completes an error
-func SetError(err error, format string, args ...interface{}) error {
-	msg := fmt.Errorf("%s: %v", fmt.Sprintf(format, args...), err)
-	cdsErr, ok := err.(Error)
-	if !ok {
-		return NewError(cdsErr, msg)
+func (w errorWithStack) Error() string {
+	return fmt.Sprintf("%s: %s (caused by: %s)", w.stack.String(), w.httpError, w.root)
+}
+func (w errorWithStack) Cause() error { return w.root }
+
+func (w errorWithStack) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 'v':
+		if s.Flag('+') {
+			fmt.Fprintf(s, "%+v", w.Cause())
+			return
+		}
+		fallthrough
+	case 's':
+		_, _ = io.WriteString(s, w.Error())
 	}
-	return msg
 }
 
-// ProcessError tries to recognize given error and return error message in a language matching Accepted-Language
-func ProcessError(target error, al string) (string, Error) {
-	// will recursively retrieve the topmost error which does not implement causer, which is assumed to be the original cause
-	target = errors.Cause(target)
-	cdsErr, ok := target.(Error)
-	if !ok || cdsErr.ID == 0 {
-		if ok && cdsErr.Message != "" {
-			return cdsErr.Message, cdsErr
+// IsErrorWithStack returns true if given error is an errorWithStack.
+func IsErrorWithStack(err error) bool {
+	_, ok := err.(errorWithStack)
+	return ok
+}
+
+type stack []uintptr
+
+func (s *stack) String() string {
+	var names []string
+	for _, pc := range *s {
+		name := runtime.FuncForPC(pc).Name()
+		if strings.HasPrefix(name, "github.com/ovh/cds/vendor") {
+			continue
+		}
+		if strings.HasPrefix(name, "github.com/ovh/cds") {
+			sp := strings.Split(name, "/")
+			sp = strings.Split(sp[len(sp)-1], ".")
+			// check if it's a struct or package func
+			if strings.HasPrefix(sp[1], "(") {
+				names = append(names, sp[2])
+			} else {
+				names = append(names, sp[1])
+			}
 		}
-		return errorsAmericanEnglish[ErrUnknownError.ID], ErrUnknownError
 	}
-	acceptedLanguages, _, err := language.ParseAcceptLanguage(al)
+	reverse(names)
+	return strings.Join(names, ">")
+}
+
+func reverse(ss []string) {
```

Comment:
maybe usage of https://golang.org/pkg/sort/#example_Reverse is better ? Don't know

Reply:
sort.Reverse is used to reverse a sort.